### PR TITLE
Add some generic database types

### PR DESCRIPTION
**Pull request by supermari0**

### CODE REVIEW

File: src/Lib.hs [hunk 1]
```diff
@@ -11,27 +11,37 @@ someFunc :: IO ()
...
+  -- out how to do a default implementation, ideally so anything deriving the
+  -- typeclass can use it.
+  fromGenericRow (GenericRow[dcID, abbr, qID]) = DatacenterRow
+    <$> HDBC.safeFromSql dcID
+    <*> HDBC.safeFromSql abbr
+    <*> HDBC.safeFromSql qID
+  fromGenericRow r = Convertible.convError "Could not parse a datacenter row" r
+
+data DatacenterTable = DatacenterTable [ DatacenterRow ] deriving Show
+instance DatabaseTable DatacenterTable where
```
Comment (rebeccaskinner) [inline src/Lib.hs]: you don't need the `where` clause here since there aren't any functions in the type class

File: src/Lib.hs [hunk 2]
```diff
@@ -11,27 +11,37 @@ someFunc :: IO ()
...
+
+-- A generic row having any number of columns
+data GenericRow = GenericRow [ HDBC.SqlValue ] deriving Show
```
Comment (rebeccaskinner) [inline src/Lib.hs]: It might make sense to make this a `newtype` rather than `data` since `newtype` is a zero-cost abstraction.

File: src/Lib.hs [hunk 3]
```diff
@@ -11,27 +11,37 @@ someFunc :: IO ()
...
 
--- TODO make this a generic fn elevating a Row to a specific type of Row
-toDatacenterRow :: Row -> Convertible.ConvertResult DatacenterRow
-toDatacenterRow (Row [dcID, abbr, qID]) = DatacenterRow
-  <$> HDBC.safeFromSql dcID
-  <*> HDBC.safeFromSql abbr
-  <*> HDBC.safeFromSql qID
-toDatacenterRow r = Convertible.convError "Could not parse a datacenter row" r
+hdbcToGenericRow :: Maybe [HDBC.SqlValue] -> Maybe GenericRow
+hdbcToGenericRow r = GenericRow <$> r
```
Comment (rebeccaskinner) [inline src/Lib.hs]: `hdbcToGenericRow = fmap GenericRow`

**Review by rebeccaskinner — CHANGES_REQUESTED** [review]
nits

File: src/Lib.hs (hunk 1 above)
Comment (supermari0) [reply src/Lib.hs]: fixed, for some reason it wasn't building but I think that was another issue

File: src/Lib.hs (hunk 2 above)
Comment (supermari0) [reply src/Lib.hs]: done

Comment (supermari0) [reply src/Lib.hs]: also done for DatacenterTable

File: src/Lib.hs (hunk 3 above)
Comment (supermari0) [reply src/Lib.hs]: done

**Review by rebeccaskinner — APPROVED** [review]
lgtm

**Pull request merged**